### ISSUE TITLE
custom route parameters?

### ISSUE THREAD

**Issue by jackttcoms** [tumi-soft/okami#2]
how would custom route params work? such as /article/:num to allow for an article id etc. thanks

**Comment by cpt-erwin** [comment]
> how would custom route params work? such as /article/:num to allow for an article id etc. thanks

I like the approach of some frameworks where you have optional and mandatory params that I want to implement here as well.

For a mandatory route argument, it would be something like
```php
  $app->router->get('/posts/{id}', PostController::class);
```
where if you would go to route `/posts` you'll get 404 and only routes like `/posts/1` will work.

On the other hand for an optional route argument, it would be something like
```php
  $app->router->get('/posts[/{id}]', PostController::class);
```
where both `/posts` and  `/posts/1` routes will work with additional logic in the Controller checking the optional param behave accordingly.

You could also compose multiple optional params in a way like `/route[/{param_1}[/{param_2}]]` and so on.

I would also like to add the possibility to specify a regular expression for the params so if you have param like `{id}` for which you would like to only accept a number like `^[0-9]+$` you could pass that regular expression with the param like `{id:^[0-9]+$}`.

Or maybe I'll stick with strict types and you'll be able to set the type of the param like `{id:int}` instead of the regular expression.

I hope this will show you the basic idea I have for now.

I'm open to any suggestion and I'll be more than happy to have a conversation about this and other topics as well. 😉

**Comment by jackttcoms** [comment]
Yes that is my thinking too. Sounds amazing and would be great for that to be implemented especially the params and mulitple params.

Something like this:
```
    public $patterns = [
        ':any' => '.*', // Any
        ':num' => '[0-9]+', // Numbers
        ':alpha' => '[a-zA-Z]+', //Letters
        ':alnum' => '[a-zA-Z0-9]+', //Letters & numbers
    ];
```

but the strict type idea is also super cool 👍 

i also feel an `$app->router->any()` would be a helpful method :)
Hopefully this can be implemented soon, would be amazing! :)

**Comment by jackttcoms** [comment]
Also a cool implementation to easily enable basic REST API-like functionality would be with the rendering, add an option to return the data in json format rather than loading a view etc. :)

and eventually an almost middleware functionality but at the core which for example could have CSRF support when doing request methods to add additional security and prevent code duplication in the long run

Sorry for spamming you with features, just so cool :)

**Comment by cpt-erwin** [comment]
Tbh I like the way you're hyped about this project as I am!

If you have more ideas you can split them to separate issues so we can discuss them and then implement them. 😄

**Comment by jackttcoms** [comment]
:) always. love creating solutions for things :)

Ye i'll do that then can begin implementing :)

**Comment by cpt-erwin** [comment]
So I've finally started working on this issue and I wanted to know your POV @jackttcoms.

```
/
/posts
/posts/{id}
/posts/{id:id}
/posts[/{id:id}]
/posts[/{slug:slug}]
/gallery/{galleryID:id}/image/{imageID:id}
```
Those are some examples of paths that can be used in the app.

I want to come up with all cases so the logic can be built as flexible as possible.

Any other ideas or are those paths all that we need to know to work with?

**Comment by cpt-erwin** [comment]
Wildcard routing with custom params implemented in commit d362c7f8783cbf96db651f529d1544e8ac1b86a9! 🥳

There are now three ways to set up a new route in your app:

**1. Simple template rendering**
```php
$router->get('/about', 'about');
```

Visiting the URL `/about` will result in rendering the content of `about.phtml` layout inside the `views` folder.

**2. Custom function (wildcard route)**
```php
$router->get('/test/{param:alnum}[/id/{id:id}]', function(\Okami\Core\Request $request, \Okami\Core\Response $response, array $params) {
    echo '<pre>' . var_export($params, true) . '</pre>';
});
```

Using wildcard route like this will result in creating multiple routes in the background:
- `/test/([a-zA-Z0-9]+)/id/([0-9]+)`
- `/test/([a-zA-Z0-9]+)`

The `$params` array will contain keys corresponding to the naming convention used in wildcard route. For the optional route will both of the parameters be initialized with the value past in the URL. In the second case, the absence of the id parameter will result in setting the default `null` value for the `id` key in the `$params` array.

**3. Calling a specified controller (wildcard route)**
```php
$router->get('/posts/{id:id}', [PostController::class, 'detail']);
```

Visiting the appropriate page will pass the control to the specified controller by calling its method. Both the controller and the method are specified in the array as a second argument of the `get()` method.

---------------
Even though it's a very basic implementation it's overall sufficient atm. Further enhancement can be expected in the future based on my needs or your requests. 🙂
